Signal/src/call/CallService.swift

```diff
@@ -300,32 +305,49 @@ protocol CallServiceObserver: class {
 
             return self.peerConnectionClient!.createOffer()
         }.then { (sessionDescription: HardenedRTCSessionDescription) -> Promise<Void> in
+            guard self.call == call else {
+                return Promise(error: CallError.obsoleteCall(description:"obsolete call in \(#function)"))
+            }
+
             return self.peerConnectionClient!.setLocalSessionDescription(sessionDescription).then {
                 let offerMessage = OWSCallOfferMessage(callId: call.signalingId, sessionDescription: sessionDescription.sdp)
                 let callMessage = OWSOutgoingCallMessage(thread: thread, offerMessage: offerMessage)
                 return self.messageSender.sendCallMessage(callMessage)
             }
         }.then {
+            guard self.call == call else {
+                return Promise(error: CallError.obsoleteCall(description:"obsolete call in \(#function)"))
+            }
+
             let (callConnectedPromise, fulfill, _) = Promise<Void>.pending()
             self.fulfillCallConnectedPromise = fulfill
 
             // Don't let the outgoing call ring forever. We don't support inbound ringing forever anyway.
             let timeout: Promise<Void> = after(interval: TimeInterval(connectingTimeoutSeconds)).then { () -> Void in
                 // rejecting a promise by throwing is safely a no-op if the promise has already been fulfilled
-                throw CallError.timeout(description: "timed out waiting to receive call answer", call: call)
+                throw CallError.timeout(description: "timed out waiting to receive call answer")
             }
 
             return race(timeout, callConnectedPromise)
         }.then {
+            guard self.call == call else {
+                throw CallError.obsoleteCall(description: "obsolete outgoing call connected")
+            }
+
             Logger.info("\(self.TAG) outgoing call connected.")
+
+            // TODO: There must be a better way to return a fulfilled promise.
+            let (promise, fulfill, _) = Promise<Void>.pending()
+            fulfill()
+            return promise
         }.catch { error in
             Logger.error("\(self.TAG) placing call failed with error: \(error)")
 
             if let callError = error as? CallError {
-                self.handleFailedCall(error: callError)
+                self.handleFailedCall(failedCall: call, error: callError)
```

Comment:
`handleFailedCall()` is now more careful to avoid terminating the current call in response to errors for obsolete calls, so all of its callers must pass the call associated with the error.